base.py

```diff
@@ -69,3 +76,41 @@ def _validate_input(self, X: List):
 
         X = [MolFromSmiles(x) if type(x) == str else x for x in X]
         return X
+
+
+class FingerprintGeneratorInterface(ABC):
+    def __init__(self, result_vector_type: str = "bit"):
+        """
+        result_vector_tape has to be one of the following:
+        bit, spares, count, sparse_count
+        """
+        assert result_vector_type in legal_result_types
+        self.result_vector_type = result_vector_type
+        self.fp_generator_kwargs = {}
+        if result_vector_type in ["sparse", "sparse_count"]:
+            self.concatenate_function = vstack
```

Comment:
This is not a good idea. Note that we currently have:
1. Class inherits from `FingerprintTransformer` and `FingerprintGeneratorInterface`.
2. At the beginning of inheriting class, we call parent class constructors in order.
3. `FingerprintTransformer` creates `self.concatenate_function` attribute and assigns default value `np.concatenate`
4. Then, in the line immediately below, `FingerprintGeneratorInterface` constructor is called and it overrides this value

What's more, those two parent classes do not know that other class also uses this attribute! This is a very implicit relationship, and very complicated.

I think we should try to keep to one class, `FingerprintTransformer`. Currently `FingerprintGeneratorInterface` serves 2 purposes (which also breaks Single Responsibility Principle for a mixin, BTW):
1. Provide a way to select concatenation method (`np.concatenate` vs `scipy.sparse.vstack`)
2. Use generator for calculating fingerprints

For concatenating the results, we can just do a simple if/else check at the end, like `if isinstance(results[0], np.ndarray)...`.

To calculate fingerprints with a generator, create a method in `FingerprintTransformer` and just call it in `_generate_fingerprints` in inheriting classes if applicable. And if we want sparse results, we just call `csr_array` on results.

BTW, you don't need a list and NumPy array for generators, you can use `GetFingerprintAsNumPy` and `GetCountFingerprintAsNumPy`.

Reply:
The concat function has been removed and it's now only using the type check as suggested( *For concatenating the results, we can just do a simple if/else check* )

I'm not sure if I understand correctly how to merge the two classes into one. As for this pull request we haven't implemented the same result vector types in other classes(ones that don't implement generators in rdkit). To prepare for a modification has been introduced so that instead of 4 possible result types passed into FingerprintGeneratorMixin constructor we pass one argument that tells whether we want a bit or count fingerprint and other argument - sparse - bool value that determines if the result is supposed to be a sparse fingerprint or not

Using GetCountFingerprintAsNumpy will help us because it will allow removal of ToList() call but we still have to create a list of numpy fingerprints as both of mentioned methodes can't take whole numpy arrays of molecules. We're only able to pass individual molecules to the function